### ISSUE TITLE
ReferenceArrayInput no longer works as a valid filter component

### ISSUE THREAD

**Issue by andrico1234** [marmelab/react-admin#7470]
In RA v3, the following component was a valid filter 

```javascript
const postFilter = [
  <ReferenceArrayInput
    alwaysOn
    label="Comments"
    source="commentId"
    reference="comments"
    filterToQuery={(query: string) => ({ fullName: query })}
    allowEmpty={false}
  >
    <AutocompleteArrayInput placeholder="e.g. Jane Doe" source="id" />
  </ReferenceArrayInput>
];
```

I would expect this to retrieve the record from comments using the query as a filter.

Doing so now displays the following error:

```
i.map is not a function


    at ReferenceArrayInput (https://izocqy.sse.codesandbox.io/index.bundle.js:93899:101570)
    at div
    at https://izocqy.sse.codesandbox.io/index.bundle.js:590:66
    at FilterFormInput (https://izocqy.sse.codesandbox.io/index.bundle.js:93899:42088)
    at form
```

I've added a [repro case here](https://codesandbox.io/s/vigilant-cloud-izocqy?file=/src/posts/PostList.tsx:457-769)

If you go to the PostList page, you can see that the page breaks. 

I haven't done any digging myself, but my main guess is that the `AutocompleteArrayInput` component doesn't handle the loading state correctly, or when there's no data currently available. Not sure if this is the likely problem.

When run on my local device, I get the following snippet:

```
// useGetManyAggregate

 74 | const queryCache = queryClient.getQueryCache();
  75 | const { ids, meta } = params;
  76 | const placeholderData = useMemo(() => {
> 77 |     const records = ids.map(id => {
     | ^  78 |         const queryHash = hashQueryKey([
  79 |             resource,
  80 |             'getOne',
```

I've also done the following and am still getting the same issue:

```
function CustomFilter() {
return (
    <Filter {...props}>
      <ReferenceArrayInput
        alwaysOn
        label="Comment name"
        source="commentId"
        reference="comments"
        filterToQuery={(query: string) => ({ fullName: query })}
        allowEmpty={false}
      >
        <TempArrayInput
          placeholder="e.g. Jane Doe"
          source="id"
          shouldRenderSuggestions={(query: string) => query.length > 1}
          optionText="attributes.fullName"
        />
      </ReferenceArrayInput>
   )}

function TempArrayInput(props: any) {
  return null;
}

```

This leads me to believe that the problem is with the `ReferenceArrayInput` and not the child component.

**Other information:**

**Environment**

* React-admin version: 4.0.0.rc-0
* Last version that did not exhibit the issue (if applicable):
* React version:
* Browser:
* Stack trace (in case of a JS error):

**Comment by WiXSL** [comment]
Reproduced. Thanks for the report

**Comment by fzaninotto** [comment]
It seems react-hook-form's `useWatch` returns an object instead of an array for the value if the related field. 

When I test with the following filter:

```jsx
const postFilter = [
    <SearchInput source="q" alwaysOn />,
    <TextInput source="title" defaultValue="Qui tempore rerum et voluptates" />,
    <QuickFilter
        label="resources.posts.fields.commentable"
        source="commentable"
        defaultValue
    />,
    <ReferenceArrayInput alwaysOn label="tags" source="tags" reference="tags">
        <SelectArrayInput optionText="name.en" />
    </ReferenceArrayInput>,
];
```

The list properly displays the empty filter, but as soon as I enter a new filter, the error message appears.

Upon debugging, it seems that `<ReferenceArrayInput source="tags">` sometimes receives `[18]` (correct value) and sometimes `{ 0: 18 }` (incorrect value). 

there must be something wronf in the `<FilterForm>` initialization.

**Comment by fzaninotto** [comment]
Confirmed: the problem is in `FilterForm`'s `mergeInitialValuesWithDefaultValues`

**Comment by fzaninotto** [comment]
Fixed by #7479
